R/cube-result.R

```diff
@@ -13,6 +13,130 @@ setMethod("initialize", "CrunchCube", function (.Object, ...) {
     return(.Object)
 })
 
+#' @rdname cube-methods
+#' @export
+
+setMethod("[", "CrunchCube", function (x, i, j, ..., drop = TRUE) {
+    subset <- eval(substitute(alist(i, j, ...)))
+    subset <- replaceMissingWithTRUE(subset)
+    translated_subset <- translateCubeIndex(x, subset, drop)
+
+    out <- x
+    out@arrays$count <- subsetByList(out@arrays$count, translated_subset, drop)
+    out@arrays$.unweighted_counts <- subsetByList(out@arrays$.unweighted_counts,
+        translated_subset, drop)
+
+    out@dims[] <- mapply(subsetArrayDimension,
+        dim = x@dims,
+        idx = translated_subset,
+        SIMPLIFY = FALSE)
+
+    if (drop) {
+        keep_args <- vapply(translated_subset, function(out) {
+            length(out) != 1 || isTRUE(out)}, FUN.VALUE = logical(1))
+        keep_dims <- names(dimnames(x@arrays$count))[keep_args]
+        out@dims <- out@dims[(names(out@dims) %in% keep_dims)]
+    }
+    return(out)
+})
+
+#' Replace missing elements with TRUE
+#'
+#' When subsetting missingness stands in for selecting all of the elements of that
+#' dimension. This gets tricky when you are selecting n dimensions and so need to
+#' capture the dimensions with `...`.  This function checks if an element of a list
+#' is missing and replaces that element with `TRUE` in order to handle this case.
+#'
+#' @param l a list
+#' @return a list
+replaceMissingWithTRUE <- function(l){
+    out <- lapply(l, function(x){
+        if (is.symbol(x)) {
+            x <- tryCatch(eval(x), error = function(c){
+                msg <- conditionMessage(c)
+                if (msg == "argument is missing, with no default") {
+                    return(TRUE)
+                } else {
+                    stop(c)
+                }
+            })
+        }
+        return(eval(x))
+    })
+    return(out)
+}
+
+subsetByList <- function(arr, arglist, drop){
+    #suppresses "named arguments other than drop are discouraged" warning
+    suppressWarnings(
+        do.call('[', c(list(x = arr, drop = drop), arglist))
+        )
+}
+
+subsetArrayDimension <- function(dim, idx){
+    dim$any.or.none <- dim$any.or.none[idx]
+    dim$missing <- dim$missing[idx]
+    dim$name <- dim$name[idx]
+    return(dim)
+}
+
+#' Translate user facing cube subset to programmatic cube subset
+#'
+#' Cubes that include multiple response variables create a special kind of complexity.
+#' Multiple response variables are actually 2d arrays with the selections along one dimension (`cat`, `dog`, `fish``)
```

Comment:
The terminology for the dimension with `cat`, `dog`, `fish` is not super standard, but Mike and I have been trying to standardise on "items", but we should definitely not call them selections (since that's generally what the _other_ dimension is referred to as)

Additionally, there's an extra \` after `fish`

Reply:
"responses" is the other word we use, which calls back to variable type name.